### ISSUE TITLE
Pending threads after headless execution

### ISSUE THREAD

**Issue by romainreuillon** [NetLogo/NetLogo#876]
OpenMOLE calls the headless version of netlogo 5.2: https://github.com/openmole/openmole/blob/master/openmole/plugins/org.openmole.plugin.task.netlogo/src/main/scala/org/openmole/plugin/task/netlogo/NetLogoTask.scala

It needs to execute many executions of several netlogo models in various threads. The netlogo headless workspace is created using:
```java
    protected HeadlessWorkspace workspace = HeadlessWorkspace.newInstance();

    @Override
    public void open(String script) throws Exception {
        workspace.open(script);
    }
```

At the end of the execution the workspace is disposed:
```java
    @Override
    public void dispose() throws Exception {
        workspace.dispose();
    }
```

However after many execution there are thousands of threads like that, remaining in the JVM:
```
"JobThread" #41890 daemon prio=4 os_prio=0 tid=0x00007f21ee2d5800 nid=0xc2a2 in Object.wait() [0x00007f1d3bdcf000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
        at java.lang.Object.wait(Native Method)
        at org.nlogo.job.JobThread$$anonfun$run$1$$anonfun$apply$mcV$sp$1.apply$mcV$sp(JobThread.scala:87)
        at org.nlogo.job.JobThread$$anonfun$run$1$$anonfun$apply$mcV$sp$1.apply(JobThread.scala:87)
        at org.nlogo.job.JobThread$$anonfun$run$1$$anonfun$apply$mcV$sp$1.apply(JobThread.scala:87)
        at scala.util.control.Exception$Catch.apply(Exception.scala:88)
        at org.nlogo.util.Exceptions$.ignoring(Exceptions.scala:38)
        at org.nlogo.job.JobThread$$anonfun$run$1.apply$mcV$sp(JobThread.scala:84)
        - locked <0x00000000e7eb44d0> (a java.lang.Object)
        at org.nlogo.job.JobThread$$anonfun$run$1.apply(JobThread.scala:76)
        at org.nlogo.job.JobThread$$anonfun$run$1.apply(JobThread.scala:76)
        at scala.util.control.Exception$Catch.apply(Exception.scala:88)
        at org.nlogo.util.Exceptions$.handling(Exceptions.scala:41)
        at org.nlogo.job.JobThread.run(JobThread.scala:75) 
```

How would it be possible to get a clean state after a NetLogo execution?

Cheers,
Romain

**Comment by SethTisue** [comment]
In our own testing, `dispose()` does not leave JobThreads lying around like this — at least after `System.gc()` has run.

Can you isolate the conditions under which the problem occurs? For example, does it happen only with particular models, or with any model, even a simple one such as Fire or Wolf-Sheep Predation?

**Comment by romainreuillon** [comment]
This happened after executing ants many times. I just examined the stack and did not investigated it further, for now...

**Comment by mrerrormessage** [comment]
Are you using extensions? There are some known bugs around the `nw` extension creating lots of extra threads (though not job threads), but it's possible that other extensions might also create threads.

**Comment by romainreuillon** [comment]
Here is the stack after executing the Ants models a few times in lplain netlogo (no plugin).

OpenMOLE uses OSGi, so we have to change the thread classloader to for NetLogo to find some resources (it would be better if it would be performed through the class classloader, but it would be another issue). So we encapsulate the calls to, the command method of the headless workspace with:
```scala
  private def withClassLoader[T](classLoader: ClassLoader)(f: ⇒ T): T = {
    val threadClassLoader = Thread.currentThread().getContextClassLoader
    Thread.currentThread().setContextClassLoader(netLogoFactory().getNetLogoClassLoader)
    try f
    finally Thread.currentThread().setContextClassLoader(threadClassLoader)
  }
```

Could it play a role in this thread leak?

**Comment by romainreuillon** [comment]
The file content:
```
2015-11-03 16:53:10
Full thread dump OpenJDK 64-Bit Server VM (25.66-b17 mixed mode):

"JobThread" #63 daemon prio=4 os_prio=0 tid=0x00007f97e8041800 nid=0x2675 in Object.wait() [0x00007f979d21b000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	at org.nlogo.job.JobThread$$anonfun$run$1$$anonfun$apply$mcV$sp$1.apply$mcV$sp(JobThread.scala:87)
	at org.nlogo.job.JobThread$$anonfun$run$1$$anonfun$apply$mcV$sp$1.apply(JobThread.scala:87)
	at org.nlogo.job.JobThread$$anonfun$run$1$$anonfun$apply$mcV$sp$1.apply(JobThread.scala:87)
	at scala.util.control.Exception$Catch.apply(Exception.scala:88)
	at org.nlogo.util.Exceptions$.ignoring(Exceptions.scala:38)
	at org.nlogo.job.JobThread$$anonfun$run$1.apply$mcV$sp(JobThread.scala:84)
	- locked <0x00000000d0c80208> (a java.lang.Object)
	at org.nlogo.job.JobThread$$anonfun$run$1.apply(JobThread.scala:76)
	at org.nlogo.job.JobThread$$anonfun$run$1.apply(JobThread.scala:76)
	at scala.util.control.Exception$Catch.apply(Exception.scala:88)
	at org.nlogo.util.Exceptions$.handling(Exceptions.scala:41)
	at org.nlogo.job.JobThread.run(JobThread.scala:75)

"JobThread" #61 daemon prio=4 os_prio=0 tid=0x00007f97e8092800 nid=0x2673 in Object.wait() [0x00007f979da1c000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	at org.nlogo.job.JobThread$$anonfun$run$1$$anonfun$apply$mcV$sp$1.apply$mcV$sp(JobThread.scala:87)
	at org.nlogo.job.JobThread$$anonfun$run$1$$anonfun$apply$mcV$sp$1.apply(JobThread.scala:87)
	at org.nlogo.job.JobThread$$anonfun$run$1$$anonfun$apply$mcV$sp$1.apply(JobThread.scala:87)
	at scala.util.control.Exception$Catch.apply(Exception.scala:88)
	at org.nlogo.util.Exceptions$.ignoring(Exceptions.scala:38)
	at org.nlogo.job.JobThread$$anonfun$run$1.apply$mcV$sp(JobThread.scala:84)
	- locked <0x00000000d0c808f0> (a java.lang.Object)
	at org.nlogo.job.JobThread$$anonfun$run$1.apply(JobThread.scala:76)
	at org.nlogo.job.JobThread$$anonfun$run$1.apply(JobThread.scala:76)
	at scala.util.control.Exception$Catch.apply(Exception.scala:88)
	at org.nlogo.util.Exceptions$.handling(Exceptions.scala:41)
	at org.nlogo.job.JobThread.run(JobThread.scala:75)

"JobThread" #59 daemon prio=4 os_prio=0 tid=0x00007f97e839e800 nid=0x2670 in Object.wait() [0x00007f979e41d000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	at org.nlogo.job.JobThread$$anonfun$run$1$$anonfun$apply$mcV$sp$1.apply$mcV$sp(JobThread.scala:87)
	at org.nlogo.job.JobThread$$anonfun$run$1$$anonfun$apply$mcV$sp$1.apply(JobThread.scala:87)
	at org.nlogo.job.JobThread$$anonfun$run$1$$anonfun$apply$mcV$sp$1.apply(JobThread.scala:87)
	at scala.util.control.Exception$Catch.apply(Exception.scala:88)
	at org.nlogo.util.Exceptions$.ignoring(Exceptions.scala:38)
	at org.nlogo.job.JobThread$$anonfun$run$1.apply$mcV$sp(JobThread.scala:84)
	- locked <0x00000000d0c80b20> (a java.lang.Object)
	at org.nlogo.job.JobThread$$anonfun$run$1.apply(JobThread.scala:76)
	at org.nlogo.job.JobThread$$anonfun$run$1.apply(JobThread.scala:76)
	at scala.util.control.Exception$Catch.apply(Exception.scala:88)
	at org.nlogo.util.Exceptions$.handling(Exceptions.scala:41)
	at org.nlogo.job.JobThread.run(JobThread.scala:75)

"JobThread" #57 daemon prio=4 os_prio=0 tid=0x00007f97e8087000 nid=0x2625 in Object.wait() [0x00007f979ec1e000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	at org.nlogo.job.JobThread$$anonfun$run$1$$anonfun$apply$mcV$sp$1.apply$mcV$sp(JobThread.scala:87)
	at org.nlogo.job.JobThread$$anonfun$run$1$$anonfun$apply$mcV$sp$1.apply(JobThread.scala:87)
	at org.nlogo.job.JobThread$$anonfun$run$1$$anonfun$apply$mcV$sp$1.apply(JobThread.scala:87)
	at scala.util.control.Exception$Catch.apply(Exception.scala:88)
	at org.nlogo.util.Exceptions$.ignoring(Exceptions.scala:38)
	at org.nlogo.job.JobThread$$anonfun$run$1.apply$mcV$sp(JobThread.scala:84)
	- locked <0x00000000ccce2488> (a java.lang.Object)
	at org.nlogo.job.JobThread$$anonfun$run$1.apply(JobThread.scala:76)
	at org.nlogo.job.JobThread$$anonfun$run$1.apply(JobThread.scala:76)
	at scala.util.control.Exception$Catch.apply(Exception.scala:88)
	at org.nlogo.util.Exceptions$.handling(Exceptions.scala:41)
	at org.nlogo.job.JobThread.run(JobThread.scala:75)

"JobThread" #55 daemon prio=4 os_prio=0 tid=0x00007f97e81d9800 nid=0x260e in Object.wait() [0x00007f979f61f000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	at org.nlogo.job.JobThread$$anonfun$run$1$$anonfun$apply$mcV$sp$1.apply$mcV$sp(JobThread.scala:87)
	at org.nlogo.job.JobThread$$anonfun$run$1$$anonfun$apply$mcV$sp$1.apply(JobThread.scala:87)
	at org.nlogo.job.JobThread$$anonfun$run$1$$anonfun$apply$mcV$sp$1.apply(JobThread.scala:87)
	at scala.util.control.Exception$Catch.apply(Exception.scala:88)
	at org.nlogo.util.Exceptions$.ignoring(Exceptions.scala:38)
	at org.nlogo.job.JobThread$$anonfun$run$1.apply$mcV$sp(JobThread.scala:84)
	- locked <0x00000000ccc6b438> (a java.lang.Object)
	at org.nlogo.job.JobThread$$anonfun$run$1.apply(JobThread.scala:76)
	at org.nlogo.job.JobThread$$anonfun$run$1.apply(JobThread.scala:76)
	at scala.util.control.Exception$Catch.apply(Exception.scala:88)
	at org.nlogo.util.Exceptions$.handling(Exceptions.scala:41)
	at org.nlogo.job.JobThread.run(JobThread.scala:75)

"Timer-0" #52 daemon prio=4 os_prio=0 tid=0x00007f97bc042000 nid=0x260b in Object.wait() [0x00007f97b7bfa000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	at java.util.TimerThread.mainLoop(Timer.java:552)
	- locked <0x00000000cc503ae8> (a java.util.TaskQueue)
	at java.util.TimerThread.run(Timer.java:505)

"JobThread" #51 daemon prio=4 os_prio=0 tid=0x00007f97e823f800 nid=0x25fe in Object.wait() [0x00007f97a3216000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	at org.nlogo.job.JobThread$$anonfun$run$1$$anonfun$apply$mcV$sp$1.apply$mcV$sp(JobThread.scala:87)
	at org.nlogo.job.JobThread$$anonfun$run$1$$anonfun$apply$mcV$sp$1.apply(JobThread.scala:87)
	at org.nlogo.job.JobThread$$anonfun$run$1$$anonfun$apply$mcV$sp$1.apply(JobThread.scala:87)
	at scala.util.control.Exception$Catch.apply(Exception.scala:88)
	at org.nlogo.util.Exceptions$.ignoring(Exceptions.scala:38)
	at org.nlogo.job.JobThread$$anonfun$run$1.apply$mcV$sp(JobThread.scala:84)
	- locked <0x00000000cc503b00> (a java.lang.Object)
	at org.nlogo.job.JobThread$$anonfun$run$1.apply(JobThread.scala:76)
	at org.nlogo.job.JobThread$$anonfun$run$1.apply(JobThread.scala:76)
	at scala.util.control.Exception$Catch.apply(Exception.scala:88)
	at org.nlogo.util.Exceptions$.handling(Exceptions.scala:41)
	at org.nlogo.job.JobThread.run(JobThread.scala:75)

"Thread-20" #49 daemon prio=5 os_prio=0 tid=0x00007f98041d3800 nid=0x25fc waiting on condition [0x00007f97fd019000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000000cc87f608> (a java.util.concurrent.Semaphore$NonfairSync)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:997)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1304)
	at java.util.concurrent.Semaphore.acquire(Semaphore.java:312)
	at org.openmole.tool.collection.PriorityQueue$class.dequeue(PriorityQueue.scala:58)
	at org.openmole.tool.collection.PriorityQueue$$anon$1.dequeue(PriorityQueue.scala:28)
	at org.openmole.core.workflow.execution.local.ExecutorPool.takeNextjob(ExecutorPool.scala:54)
	at org.openmole.core.workflow.execution.local.LocalExecutor.run(LocalExecutor.scala:55)
	at java.lang.Thread.run(Thread.java:745)

"NonBlockingInputStreamThread" #48 daemon prio=5 os_prio=0 tid=0x00007f980454d000 nid=0x25fb in Object.wait() [0x00007f97a160f000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	at jline.internal.NonBlockingInputStream.run(NonBlockingInputStream.java:278)
	- locked <0x00000000c8098568> (a jline.internal.NonBlockingInputStream)
	at java.lang.Thread.run(Thread.java:745)

"Attach Listener" #47 daemon prio=9 os_prio=0 tid=0x00007f97ec001800 nid=0x25b2 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"NonBlockingInputStreamThread" #46 daemon prio=5 os_prio=0 tid=0x00007f9804041800 nid=0x2510 in Object.wait() [0x00007f97b7cfb000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	at jline.internal.NonBlockingInputStream.run(NonBlockingInputStream.java:278)
	- locked <0x00000000c52f2e60> (a jline.internal.NonBlockingInputStream)
	at java.lang.Thread.run(Thread.java:745)

"qtp236220307-45" #45 prio=5 os_prio=0 tid=0x00007f9804001000 nid=0x250e waiting on condition [0x00007f97a1d10000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000000c0500000> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
	at org.eclipse.jetty.util.BlockingArrayQueue.poll(BlockingArrayQueue.java:389)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.idleJobPoll(QueuedThreadPool.java:531)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.access$700(QueuedThreadPool.java:47)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:590)
	at java.lang.Thread.run(Thread.java:745)

"qtp236220307-43" #43 prio=5 os_prio=0 tid=0x00007f97d0003800 nid=0x250c waiting on condition [0x00007f97a1f12000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000000c0500000> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
	at org.eclipse.jetty.util.BlockingArrayQueue.poll(BlockingArrayQueue.java:389)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.idleJobPoll(QueuedThreadPool.java:531)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.access$700(QueuedThreadPool.java:47)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:590)
	at java.lang.Thread.run(Thread.java:745)

"qtp236220307-42" #42 prio=5 os_prio=0 tid=0x00007f97c4005800 nid=0x250b waiting on condition [0x00007f97a7bbb000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000000c0500000> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
	at org.eclipse.jetty.util.BlockingArrayQueue.poll(BlockingArrayQueue.java:389)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.idleJobPoll(QueuedThreadPool.java:531)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.access$700(QueuedThreadPool.java:47)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:590)
	at java.lang.Thread.run(Thread.java:745)

"NonBlockingInputStreamThread" #41 daemon prio=5 os_prio=0 tid=0x00007f9840fee000 nid=0x2509 in Object.wait() [0x00007f97a2413000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	at jline.internal.NonBlockingInputStream.run(NonBlockingInputStream.java:278)
	- locked <0x00000000c0500238> (a jline.internal.NonBlockingInputStream)
	at java.lang.Thread.run(Thread.java:745)

"Scheduler-929219836" #38 prio=5 os_prio=0 tid=0x00007f97b0001800 nid=0x24fd waiting on condition [0x00007f97a2714000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000000c0500630> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
	at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1093)
	at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:809)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1067)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1127)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

"ForkJoinPool-1-worker-5" #37 daemon prio=5 os_prio=0 tid=0x00007f9840d35800 nid=0x24f9 waiting on condition [0x00007f97a2a15000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000000c0500898> (a scala.concurrent.forkjoin.ForkJoinPool)
	at scala.concurrent.forkjoin.ForkJoinPool.scan(ForkJoinPool.java:2075)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)

"AWT-EventQueue-0" #35 prio=6 os_prio=0 tid=0x00007f9840bcd800 nid=0x24ed waiting on condition [0x00007f97a7aba000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000000c0500bb0> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at java.awt.EventQueue.getNextEvent(EventQueue.java:554)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:170)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)

"AWT-XAWT" #33 daemon prio=6 os_prio=0 tid=0x00007f9840b60000 nid=0x24e9 runnable [0x00007f97b4ed6000]
   java.lang.Thread.State: RUNNABLE
	at sun.awt.X11.XToolkit.waitForEvents(Native Method)
	at sun.awt.X11.XToolkit.run(XToolkit.java:568)
	at sun.awt.X11.XToolkit.run(XToolkit.java:532)
	at java.lang.Thread.run(Thread.java:745)

"Java2D Disposer" #31 daemon prio=10 os_prio=0 tid=0x00007f9840b3d800 nid=0x24e8 in Object.wait() [0x00007f97b53e7000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:143)
	- locked <0x00000000c05012d8> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:164)
	at sun.java2d.Disposer.run(Disposer.java:148)
	at java.lang.Thread.run(Thread.java:745)

"org.eclipse.jetty.server.session.HashSessionManager@6dc1dc69Timer" #30 daemon prio=5 os_prio=0 tid=0x00007f9840727000 nid=0x24e7 waiting on condition [0x00007f97b78f7000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000000c0501490> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
	at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1093)
	at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:809)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1067)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1127)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

"qtp236220307-29" #29 prio=5 os_prio=0 tid=0x00007f9840503800 nid=0x24e6 waiting on condition [0x00007f97b79f8000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000000c0500000> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
	at org.eclipse.jetty.util.BlockingArrayQueue.poll(BlockingArrayQueue.java:389)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.idleJobPoll(QueuedThreadPool.java:531)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.access$700(QueuedThreadPool.java:47)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:590)
	at java.lang.Thread.run(Thread.java:745)

"qtp236220307-28" #28 prio=5 os_prio=0 tid=0x00007f98403ec800 nid=0x24e5 waiting on condition [0x00007f97b7af9000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000000c0500000> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
	at org.eclipse.jetty.util.BlockingArrayQueue.poll(BlockingArrayQueue.java:389)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.idleJobPoll(QueuedThreadPool.java:531)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.access$700(QueuedThreadPool.java:47)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:590)
	at java.lang.Thread.run(Thread.java:745)

"qtp236220307-25" #25 prio=5 os_prio=0 tid=0x00007f98404a9000 nid=0x24e2 waiting on condition [0x00007f97b7dfc000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000000c0500000> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
	at org.eclipse.jetty.util.BlockingArrayQueue.poll(BlockingArrayQueue.java:389)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.idleJobPoll(QueuedThreadPool.java:531)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.access$700(QueuedThreadPool.java:47)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:590)
	at java.lang.Thread.run(Thread.java:745)

"qtp236220307-24-acceptor-0@491342b0-ServerConnector@47d4b5e2{SSL-HTTP/1.1}{localhost:46447}" #24 prio=5 os_prio=0 tid=0x00007f98404a7000 nid=0x24e1 runnable [0x00007f97b7efd000]
   java.lang.Thread.State: RUNNABLE
	at sun.nio.ch.ServerSocketChannelImpl.accept0(Native Method)
	at sun.nio.ch.ServerSocketChannelImpl.accept(ServerSocketChannelImpl.java:422)
	at sun.nio.ch.ServerSocketChannelImpl.accept(ServerSocketChannelImpl.java:250)
	- locked <0x00000000c0501d50> (a java.lang.Object)
	at org.eclipse.jetty.server.ServerConnector.accept(ServerConnector.java:377)
	at org.eclipse.jetty.server.AbstractConnector$Acceptor.run(AbstractConnector.java:500)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:745)

"qtp236220307-23-selector-ServerConnectorManager@550de6a9/1" #23 prio=5 os_prio=0 tid=0x00007f98405d3800 nid=0x24e0 runnable [0x00007f97b7ffe000]
   java.lang.Thread.State: RUNNABLE
	at sun.nio.ch.EPollArrayWrapper.epollWait(Native Method)
	at sun.nio.ch.EPollArrayWrapper.poll(EPollArrayWrapper.java:269)
	at sun.nio.ch.EPollSelectorImpl.doSelect(EPollSelectorImpl.java:79)
	at sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:86)
	- locked <0x00000000c0502038> (a sun.nio.ch.Util$2)
	- locked <0x00000000c0502048> (a java.util.Collections$UnmodifiableSet)
	- locked <0x00000000c0501ff0> (a sun.nio.ch.EPollSelectorImpl)
	at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:97)
	at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:101)
	at org.eclipse.jetty.io.SelectorManager$ManagedSelector.select(SelectorManager.java:596)
	at org.eclipse.jetty.io.SelectorManager$ManagedSelector.run(SelectorManager.java:545)
	at org.eclipse.jetty.util.thread.NonBlockingThread.run(NonBlockingThread.java:52)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:745)

"qtp236220307-22-selector-ServerConnectorManager@550de6a9/0" #22 prio=5 os_prio=0 tid=0x00007f984065b000 nid=0x24df runnable [0x00007f97fc17b000]
   java.lang.Thread.State: RUNNABLE
	at sun.nio.ch.EPollArrayWrapper.epollWait(Native Method)
	at sun.nio.ch.EPollArrayWrapper.poll(EPollArrayWrapper.java:269)
	at sun.nio.ch.EPollSelectorImpl.doSelect(EPollSelectorImpl.java:79)
	at sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:86)
	- locked <0x00000000c05022d8> (a sun.nio.ch.Util$2)
	- locked <0x00000000c05022e8> (a java.util.Collections$UnmodifiableSet)
	- locked <0x00000000c0502290> (a sun.nio.ch.EPollSelectorImpl)
	at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:97)
	at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:101)
	at org.eclipse.jetty.io.SelectorManager$ManagedSelector.select(SelectorManager.java:596)
	at org.eclipse.jetty.io.SelectorManager$ManagedSelector.run(SelectorManager.java:545)
	at org.eclipse.jetty.util.thread.NonBlockingThread.run(NonBlockingThread.java:52)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:745)

"Thread-3" #20 daemon prio=5 os_prio=0 tid=0x00007f9840424800 nid=0x24dd waiting on condition [0x00007f97fceb7000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000000c05024e8> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
	at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1093)
	at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:809)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1067)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1127)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

"Bundle File Closer" #19 daemon prio=5 os_prio=0 tid=0x00007f9840a90800 nid=0x24d5 in Object.wait() [0x00007f97fcd9e000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	at java.lang.Object.wait(Object.java:502)
	at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.getNextEvent(EventManager.java:400)
	- locked <0x00000000c0c1fb70> (a org.eclipse.osgi.framework.eventmgr.EventManager$EventThread)
	at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.run(EventManager.java:336)

"Worker-JM" #16 daemon prio=5 os_prio=0 tid=0x00007f98407e9800 nid=0x24d4 in Object.wait() [0x00007f97fdc18000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000000c0c1fe78> (a java.util.ArrayList)
	at org.eclipse.core.internal.jobs.InternalWorker.run(InternalWorker.java:59)
	- locked <0x00000000c0c1fe78> (a java.util.ArrayList)

"Refresh Thread: Equinox Container: 80ef1691-4282-0015-14e8-f9559e5c64d2" #15 daemon prio=5 os_prio=0 tid=0x00007f98407ca000 nid=0x24d3 in Object.wait() [0x00007f97fdf19000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000000c0c1fe90> (a org.eclipse.osgi.framework.eventmgr.EventManager$EventThread)
	at java.lang.Object.wait(Object.java:502)
	at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.getNextEvent(EventManager.java:400)
	- locked <0x00000000c0c1fe90> (a org.eclipse.osgi.framework.eventmgr.EventManager$EventThread)
	at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.run(EventManager.java:336)

"Start Level: Equinox Container: 80ef1691-4282-0015-14e8-f9559e5c64d2" #14 daemon prio=5 os_prio=0 tid=0x00007f98403b8800 nid=0x24d2 in Object.wait() [0x00007f97fe586000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000000c0981750> (a org.eclipse.osgi.framework.eventmgr.EventManager$EventThread)
	at java.lang.Object.wait(Object.java:502)
	at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.getNextEvent(EventManager.java:400)
	- locked <0x00000000c0981750> (a org.eclipse.osgi.framework.eventmgr.EventManager$EventThread)
	at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.run(EventManager.java:336)

"Framework Event Dispatcher: Equinox Container: 80ef1691-4282-0015-14e8-f9559e5c64d2" #13 daemon prio=5 os_prio=0 tid=0x00007f98403b6000 nid=0x24d1 in Object.wait() [0x00007f97fe687000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	at java.lang.Object.wait(Object.java:502)
	at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.getNextEvent(EventManager.java:400)
	- locked <0x00000000c09818d8> (a org.eclipse.osgi.framework.eventmgr.EventManager$EventThread)
	at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.run(EventManager.java:336)

"Active Thread: Equinox Container: 80ef1691-4282-0015-14e8-f9559e5c64d2" #11 prio=5 os_prio=0 tid=0x00007f984038e000 nid=0x24d0 waiting on condition [0x00007f97fe788000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000000c0981bd8> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
	at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1093)
	at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:809)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1067)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1127)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

"Service Thread" #9 daemon prio=9 os_prio=0 tid=0x00007f984017b000 nid=0x24ce runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C1 CompilerThread2" #8 daemon prio=9 os_prio=0 tid=0x00007f9840178000 nid=0x24cd waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C2 CompilerThread1" #7 daemon prio=9 os_prio=0 tid=0x00007f9840173800 nid=0x24cc waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C2 CompilerThread0" #6 daemon prio=9 os_prio=0 tid=0x00007f9840171800 nid=0x24cb waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Signal Dispatcher" #5 daemon prio=9 os_prio=0 tid=0x00007f984016f000 nid=0x24ca runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Surrogate Locker Thread (Concurrent GC)" #4 daemon prio=9 os_prio=0 tid=0x00007f984016d000 nid=0x24c7 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Finalizer" #3 daemon prio=8 os_prio=0 tid=0x00007f9840142800 nid=0x24c2 in Object.wait() [0x00007f97ffdfc000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:143)
	- locked <0x00000000c01a4e68> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:164)
	at java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:209)

"Reference Handler" #2 daemon prio=10 os_prio=0 tid=0x00007f9840140000 nid=0x24c1 in Object.wait() [0x00007f97ffefd000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	at java.lang.Object.wait(Object.java:502)
	at java.lang.ref.Reference$ReferenceHandler.run(Reference.java:157)
	- locked <0x00000000c01bf770> (a java.lang.ref.Reference$Lock)

"main" #1 prio=5 os_prio=0 tid=0x00007f984000b000 nid=0x24b1 waiting on condition [0x00007f98490e3000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000000c1d1b218> (a java.util.concurrent.Semaphore$NonfairSync)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:997)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1304)
	at java.util.concurrent.Semaphore.acquire(Semaphore.java:312)
	at org.openmole.gui.server.core.GUIServer.join(GUIServer.scala:105)
	at org.openmole.ui.Application$$anonfun$start$1$$anonfun$9.apply(Application.scala:202)
	at org.openmole.ui.Application$$anonfun$start$1$$anonfun$9.apply(Application.scala:190)
	at org.openmole.tool.file.FilePackage$class.withClosable(package.scala:451)
	at org.openmole.tool.file.package$.withClosable(package.scala:495)
	at org.openmole.tool.file.FilePackage$FileDecorator$$anonfun$withFileOutputStream$1.apply(package.scala:408)
	at org.openmole.tool.file.FilePackage$FileDecorator$$anonfun$withFileOutputStream$1.apply(package.scala:408)
	at org.openmole.ui.Application$$anonfun$start$1.apply(Application.scala:190)
	at org.openmole.ui.Application$$anonfun$start$1.apply(Application.scala:60)
	at org.openmole.core.replication.DBServerRunning$.useDB(DBServerRunning.scala:42)
	at org.openmole.ui.Application.start(Application.scala:60)
	at org.openmole.ui.Application.start(Application.scala:47)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:196)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:134)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:382)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:236)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:629)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:584)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1438)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1414)

"VM Thread" os_prio=0 tid=0x00007f984013b000 nid=0x24c0 runnable 

"Gang worker#0 (Parallel GC Threads)" os_prio=0 tid=0x00007f9840020000 nid=0x24b2 runnable 

"Gang worker#1 (Parallel GC Threads)" os_prio=0 tid=0x00007f9840022000 nid=0x24b3 runnable 

"Gang worker#2 (Parallel GC Threads)" os_prio=0 tid=0x00007f9840023800 nid=0x24b4 runnable 

"Gang worker#3 (Parallel GC Threads)" os_prio=0 tid=0x00007f9840025800 nid=0x24b5 runnable 

"G1 Main Concurrent Mark GC Thread" os_prio=0 tid=0x00007f9840040800 nid=0x24bb runnable 

"Gang worker#0 (G1 Parallel Marking Threads)" os_prio=0 tid=0x00007f9840042800 nid=0x24bc runnable 

"G1 Concurrent Refinement Thread#0" os_prio=0 tid=0x00007f984002f800 nid=0x24ba runnable 

"G1 Concurrent Refinement Thread#1" os_prio=0 tid=0x00007f984002d800 nid=0x24b9 runnable 

"G1 Concurrent Refinement Thread#2" os_prio=0 tid=0x00007f984002b800 nid=0x24b8 runnable 

"G1 Concurrent Refinement Thread#3" os_prio=0 tid=0x00007f9840029800 nid=0x24b7 runnable 

"G1 Concurrent Refinement Thread#4" os_prio=0 tid=0x00007f9840028000 nid=0x24b6 runnable 

"VM Periodic Task Thread" os_prio=0 tid=0x00007f984017e000 nid=0x24cf waiting on condition 

JNI global references: 502
```

**Comment by romainreuillon** [comment]
All model execution are ended at the point I captured this stack.

**Comment by SethTisue** [comment]
If the problem was that threads were shut down but associated objects weren't getting GC'ed afterwards, I might suspect the classloader thing. But the threads aren't even getting shut down, so that seems like something else. I don't see how that's even possible if `dispose` is being called. How certain are you that you're actually calling `dispose`?

**Comment by romainreuillon** [comment]
I just tried with a single execution and added println arround the `dispose function. Dispose is called and the leaked thread is present in the stack list.

**Comment by SethTisue** [comment]
here's a transcript showing the bug not occurring:

```text
% echo $JAVA_HOME
/Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Home
% pwd
/Applications/NetLogo 5.2.1
% /usr/local/scala-2.9.3/bin/scala -Yrepl-sync -classpath NetLogo.jar
Welcome to Scala version 2.9.3 (Java HotSpot(TM) 64-Bit Server VM, Java 1.6.0_65).
Type in expressions to have them evaluated.
Type :help for more information.

scala> val ws = org.nlogo.headless.HeadlessWorkspace.newInstance
ws: org.nlogo.headless.HeadlessWorkspace = org.nlogo.headless.HeadlessWorkspace@2c808512

scala> ws.open("/Applications/NetLogo 5.2.1/models/Sample Models/Biology/Ants.nlogo")

scala> ws.command("setup  print count turtles")
125

scala> val jm = ws.jobManager.asInstanceOf[org.nlogo.job.JobManager]
jm: org.nlogo.job.JobManager = org.nlogo.job.JobManager@1bebb6f2

scala> val field = jm.getClass.getDeclaredField("thread")
field: java.lang.reflect.Field = private final org.nlogo.job.JobThread org.nlogo.job.JobManager.thread

scala> field.setAccessible(true)

scala> val thread = field.get(jm).asInstanceOf[Thread]
thread: java.lang.Thread = Thread[JobThread,4,main]

scala> thread.isAlive
res3: Boolean = true

scala> ws.dispose()

scala> thread.isAlive
res5: Boolean = false
```

so you'll need to somehow narrow down what's different about what's going in your case, vs. what this transcript shows.

**Comment by romainreuillon** [comment]
My bad, I found the error in OpenMOLE. It was actually instantiated by mistake additional netlogo workspaces. It is now fixed:
https://github.com/openmole/openmole/commit/80f9075421d5c445a32e3bf620e0eff8d1fb61cb

Thanks a lot for your reactivity and sorry for the erroneous bug report.

**Comment by SethTisue** [comment]
np, glad I could help you narrow it down.
